src/ameba.ts

```diff
@@ -1,184 +1,123 @@
-    private taskQueue: TaskQueue = new TaskQueue();
+    private diag: DiagnosticCollection;
+    private taskQueue: TaskQueue;
     public config: AmebaConfig;
 
-    constructor(
-        diagnostics: vscode.DiagnosticCollection,
-        additionalArguments: string[] = [],
-    ) {
+    constructor(diagnostics: DiagnosticCollection) {
         this.diag = diagnostics;
-        // this.additionalArguments = additionalArguments;
+        this.taskQueue = new TaskQueue();
         this.config = getConfig();
     }
 
-    public execute(document: vscode.TextDocument, onComplete?: () => void): void {
-        if (document.languageId !== 'crystal' || document.isUntitled || !isFileUri(document.uri)) {
+    public execute(document: TextDocument): void {
+        if (document.languageId !== 'crystal' || document.isUntitled || document.uri.scheme !== 'file') {
             return;
         }
 
-        const fileName = document.fileName;
-        const uri = document.uri;
-        let currentPath = getCurrentPath(fileName);
-
-        let onDidExec = (error: Error | null, stdout: string, stderr: string) => {
-            if (this.hasError(error, stderr)) {
-                return;
-            }
+        const args = [this.config.command, document.fileName, '--format', 'json'];
+        if (this.config.configFileName.length) {
+            const dir = workspace.getWorkspaceFolder(document.uri)!.uri.fsPath;
+            args.push('--config', path.join(dir, this.config.configFileName));
+        }
 
-            this.diag.delete(uri);
-            let ameba = this.parse(stdout);
-
-            if (ameba === undefined || ameba === null) {
-                return;
-            }
-
-            let entries: [vscode.Uri, vscode.Diagnostic[]][] = [];
-            ameba.sources.forEach((source: AmebaFile) => {
-                let diagnostics: Array<vscode.Diagnostic> = [];
-                source.issues.forEach((issue: AmebaIssue) => {
-                    const loc = issue.location;
-                    let end_loc = issue.end_location;
-                    if (end_loc === null || end_loc.line === null) {
-                        end_loc = {
-                            line: loc.line,
-                            column: loc.column
-                        };
+        const task = new Task(document.uri, token => {
+            const proc = exec(args.join(' '), async (err, stdout, stderr) => {
+                if (token.isCanceled) return;
+
+                if (err && stderr.length) {
+                    if ((process.platform == 'win32' && err.code === 1) || err.code === 127) {
+                        const disable = await window.showErrorMessage(
+                            `Could not execute the Ameba file at: ${args[0]}`,
+                            'Disable (workspace)'
+                        );
+                        if (!!disable) await commands.executeCommand('crystal.ameba.disable');
+                    } else {
```

Comment:
why do we need to handle that differently on different platforms?

Reply:
The exit code for not found on Windows goes under 1 because cmd doesn't know what `ameba`/`ameba.exe`/`C:\...\ameba.exe` is, so it interprets it as an invalid function (if it interpreted it as a file then the exit code would be 2). But on Linux 1 can be anything, and 127 is specifically returned when a file/executable isn't found.